test/common/http/http2/codec_impl_test.cc

```diff
@@ -965,6 +972,134 @@ TEST_P(Http2CodecImplTest, ConnectionKeepaliveJitter) {
+
+    // Check contents for active stream, trailers to encode and header map.
+    EXPECT_THAT(
+        ostream.contents(),
+        HasSubstr(
+            "Number of active streams: 1 Active Streams:\nstream: \nConnectionImpl::StreamImpl"));
+    EXPECT_THAT(ostream.contents(),
+                HasSubstr("pending_trailers_to_encode_:   null\n"
+                          "  absl::get<RequestHeaderMapPtr>(headers_or_trailers_): \n"
+                          "    ':scheme', 'http'\n"
+                          "    ':method', 'GET'\n"
+                          "    ':authority', 'host'\n"
+                          "    ':path', '/'\n"
+                          "current_slice_: null"));
+  }
+
+  // Dump client
+  {
+    std::array<char, 2048> buffer;
+    OutputBufferStream ostream{buffer.data(), buffer.size()};
+    // Check no memory allocated.
+    Stats::TestUtil::MemoryTest memory_test;
+    client_->dumpState(ostream, 0);
+    EXPECT_EQ(memory_test.consumedBytes(), 0);
+
+    // Check contents for active stream, trailers to encode and header map.
+    EXPECT_THAT(ostream.contents(), HasSubstr("Number of active streams: 1 Active Streams:\n"
+                                              "stream: \n"
+                                              "ConnectionImpl::StreamImpl"));
+    EXPECT_THAT(ostream.contents(),
+                HasSubstr("pending_trailers_to_encode_:   null\n"
+                          "  absl::get<ResponseHeaderMapPtr>(headers_or_trailers_): \n"
+                          "    ':status', '200'\n"
+                          "current_slice_: null"));
+  }
+}
+
+TEST_P(Http2CodecImplTest, ShouldDumpCurrentSliceWithoutAllocatingMemory) {
+  initialize();
+  std::array<char, 2048> buffer;
+  OutputBufferStream ostream{buffer.data(), buffer.size()};
+  MockStreamCallbacks callbacks;
+  request_encoder_->getStream().addCallbacks(callbacks);
+
+  // Send headers
+  TestRequestHeaderMapImpl request_headers;
+  HttpTestUtility::addDefaultHeaders(request_headers);
+  TestRequestHeaderMapImpl expected_headers;
+  HttpTestUtility::addDefaultHeaders(expected_headers);
+  EXPECT_CALL(request_decoder_, decodeHeaders_(HeaderMapEqual(&expected_headers), false));
+  EXPECT_TRUE(request_encoder_->encodeHeaders(request_headers, false).ok());
+
+  // Send data payload, dump buffer as decoding data
+  EXPECT_CALL(request_decoder_, decodeData(_, false)).WillOnce(Invoke([&](Buffer::Instance&, bool) {
+    // dumpState here while we had a current slice of data. No Memory should be
+    // allocated.
+    Stats::TestUtil::MemoryTest memory_test;
+    server_->dumpState(ostream, 0);
+    EXPECT_EQ(memory_test.consumedBytes(), 0);
+  }));
+  Buffer::OwnedImpl hello("hello envoy");
+  request_encoder_->encodeData(hello, false);
+
+  // Check contents for the current slice information
+  EXPECT_THAT(ostream.contents(),
+              HasSubstr("current slice length: 20 contents: \"\0\0\\v\0\0\0\0\0\x1hello envoy"s));
```

Comment:
Make sure to adjust the \0's after you start escaping null chars.  I don't trust the code above to not treat the first \0 in this string literal as an end-of-c-string indicator

Reply:
I've now replaced this with `EndsWith` instead, that way it has to grok the remaining bytes. I needed to make some adjustments (include the newline, etc. so it makes me confident it is grokking those chars.)

See: https://en.cppreference.com/w/cpp/string/basic_string/operator%22%22s about the new C++14 string_literal.

🤞 I can appease the msvc compiler; otherwise, I'll escape as you suggest (It seems like your reasoning for escaping `\0` is tied to this?)

Reply:
The request to escape \0 goes beyond this test.  I would prefer to not have null characters in logs.

Reply:
sg, done.